### PR TITLE
Disable soft delete

### LINKED ISSUE #3: Create tests

### PR DESCRIPTION
**Pull request by apechimp**
This allows the disabling of soft delete on a per operation basis. Additionally, tests and linting have been added which resolves #3.

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -1,85 +1,88 @@
...
-            })
-            .then(function (model) {
-              return model.triggerThen('destroyed');
-            });
-        } else {
-          return mProto.destroy.apply(this, arguments);
-        }
+  var mProto = Bookshelf.Model.prototype,
+    cProto = Bookshelf.Collection.prototype;
+
```
Comment (scarney81) [inline index.js]: extra white space

File: index.js [hunk 2]
```diff
@@ -1,85 +1,88 @@
...
+
+  Bookshelf.Model = Bookshelf.Model.extend({
+    constructor: function () {
+      mProto.constructor.apply(this, arguments);
+      var self = this;
+      BPromise.resolve()
+        .then(function () {
+          if (!self.soft) {
+            throw new BPromise.CancellationError('soft delete not set');
```
Comment (scarney81) [inline index.js]: Finally! A reason to use promise cancellation

File: index.js [hunk 3]
```diff
@@ -1,85 +1,88 @@
...
+        });
+    },
+    destroy: function (opts) {
+      if (this.soft && !shouldDisable(opts)) {
+        this.set('deleted_at', new Date());
+        return this.save()
+          .tap(function (model) {
+            return model.triggerThen('destroying');
+          })
+          .then(function (model) {
```
Comment (scarney81) [inline index.js]: Any reason this is a `then` instead of a `tap`?

**Comment by scarney81** [comment]
:beers:

File: index.js (hunk 3 above)
Comment (apechimp) [inline index.js]: That's an @apolishch question. I just made whitespace changes to pass lint.

**Pull request merged**